### ISSUE TITLE
Feature Request: INT4 Quantization Support for Edge Devices

### ISSUE THREAD

**Issue by Ultraluxe25** [ultralytics/ultralytics#24860]
### Search before asking

- [x] I have searched the Ultralytics [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

First of all, I want to thank you for your amazing work. I use your models very often, and they are incredibly helpful for both my professional and personal projects.

I noticed that trained models can be easily exported using TensorRT, and it is also possible to quantize a model with INT8 to decrease memory consumption (as described here: https://docs.ultralytics.com/integrations/tensorrt#exporting-tensorrt-with-int8-quantization).

I know that many new LLMs support aggressive quantization like INT4. While INT8 is usually enough for custom fine-tuned YOLO models, I am very interested to know if the Ultralytics team plans to add INT4 quantization in future releases? This would be extremely beneficial for specific edge-device cases where minimizing the memory footprint is critical.

Thank you for your time and consideration!

### Use case

Due to the ongoing global memory market crisis and hardware constraints, there is a growing resurgence of low-VRAM GPUs (1-2 GB) and resource-constrained hardware in the consumer segment. 

Many budget enthusiasts, students, and developers working with edge devices struggle to run inference on these limited setups. While INT8 helps, implementing INT4 quantization would drastically reduce the memory requirements. This would allow users to deploy and run YOLO models in real-time on ultra-budget GPUs, micro-controllers, and edge devices that currently cannot handle even INT8. Adding INT4 would make Ultralytics models accessible to a much wider audience facing severe hardware limitations.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Ultraluxe25, thank you for your thoughtful feature request and for using Ultralytics 🚀! INT4 quantization for edge and low-memory deployment is an interesting use case, and we appreciate the clear motivation you shared. This is an automated response to help get things moving quickly, and an Ultralytics engineer will also assist soon 🤝

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).

For feature requests like this one 💡, it can also help to share any target hardware details, deployment constraints, expected runtime/export format, and whether you are considering TensorRT-based workflows so the team can better evaluate the request.

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
Hello. Thanks for the request.

Currently, we do not have plans to add INT4 quantization support because INT4 quantization is very aggressive and CNN models such as YOLO experience significant accuracy drop when quantized to INT4.

If you have any successful results from INT4 quantization for YOLO, it would be interesting to know.

**Comment by Ultraluxe25** [comment]
Got it @Y-T-G . Thanks for the detailed explanation! If I come across any successful INT4 results for YOLO, I'll share them here. Thanks!
